Bu ödevde ABD Eğitim Sistemi'ne ait bir takım verileri inceleyeceksiniz. Öncelikle veri kümesini Kaggle'dan indirin. Burada eğitim sistemine ait bir çok veri kümesi bulunmaktadır. Bu ödevde 'states_all.csv' isimli veri kümesini kullanacaksınız.

Bu ödevi tamamlamak için, aşağıdaki soruların çözümlerini içeren Jupyter notebook dosyasının Github linkini gönderin. Kafanıza takılan konuları mentörünüzle konuşabilir ya da ofis saatinde Slack üzerinden sorabilirsiniz.

- Veri kümesinde yer alan değişkenlerinin tiplerini ve her bir değişkenin eksik (null) değer oranını bulun.

- Verimizde yıl (year) sütunu olduğunu farketmişsinizdir. Şimdilik yıl verisini unutun ve her bir gözlemin aynı yıl içerisinde yapıldığını farz edin. Her bir değişken için eksik değerleri nasıl doldurabileceğinizi düşünün. Eksik değerleri bir değerle doldurmak hangi değişkenler için anlamlı, hangileri için anlamsızdır?

- Şimdi zaman faktörünü dikkate alma zamanı! 2. sorudaki cevabınızı tekrar gözden geçirin ve eksik verileri o yıl içerisinde gözlemlenen değerlere dayanarak doldurun. Örneğin, bir değeri ortalama değer ile doldurmak isterseniz, o yılın ortalamasını hesaplayın.

- Bu sefer, eksik değerleri enterpolasyon yaparak doldurun.

- İkinci, üçüncü ve dördüncü sorulardaki cevaplarınızı karşılaştırın. Anlamlı bir fark görebiliyor musunuz?

In [1]:
import numpy as np
import pandas as pd

In [2]:
df = pd.read_csv("C:/Users/Ahmet/Desktop/data/states_all.csv")
df.head()

PRIMARY_KEY       STATE  YEAR  ENROLL  TOTAL_REVENUE  FEDERAL_REVENUE  \
0     1992_ALABAMA     ALABAMA  1992     NaN      2678885.0         304177.0   
1      1992_ALASKA      ALASKA  1992     NaN      1049591.0         106780.0   
2     1992_ARIZONA     ARIZONA  1992     NaN      3258079.0         297888.0   
3    1992_ARKANSAS    ARKANSAS  1992     NaN      1711959.0         178571.0   
4  1992_CALIFORNIA  CALIFORNIA  1992     NaN     26260025.0        2072470.0   

   STATE_REVENUE  LOCAL_REVENUE  TOTAL_EXPENDITURE  INSTRUCTION_EXPENDITURE  \
0      1659028.0       715680.0          2653798.0                1481703.0   
1       720711.0       222100.0           972488.0                 498362.0   
2      1369815.0      1590376.0          3401580.0                1435908.0   
3       958785.0       574603.0          1743022.0                 964323.0   
4     16546514.0      7641041.0         27138832.0               14358922.0   

   ...  GRADES_4_G  GRADES_8_G  GRADES_12_G  GRADES_1_8_G  GRADES_9_12_G  \
0  ...     57948.0     58025.0      41167.0           NaN            NaN   
1  ...      9748.0      8789.0       6714.0           NaN            NaN   
2  ...     55433.0     49081.0      37410.0           NaN            NaN   
3  ...     34632.0     36011.0      27651.0           NaN            NaN   
4  ...    418418.0    363296.0     270675.0           NaN            NaN   

   GRADES_ALL_G  AVG_MATH_4_SCORE  AVG_MATH_8_SCORE  AVG_READING_4_SCORE  \
0      731634.0             208.0             252.0                207.0   
1      122487.0               NaN               NaN                  NaN   
2      673477.0             215.0             265.0                209.0   
3      441490.0             210.0             256.0                211.0   
4     5254844.0             208.0             261.0                202.0   

   AVG_READING_8_SCORE  
0                  NaN  
1                  NaN  
2                  NaN  
3                  NaN  
4                  NaN  

[5 rows x 25 columns]

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1715 entries, 0 to 1714
Data columns (total 25 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   PRIMARY_KEY                   1715 non-null   object 
 1   STATE                         1715 non-null   object 
 2   YEAR                          1715 non-null   int64  
 3   ENROLL                        1224 non-null   float64
 4   TOTAL_REVENUE                 1275 non-null   float64
 5   FEDERAL_REVENUE               1275 non-null   float64
 6   STATE_REVENUE                 1275 non-null   float64
 7   LOCAL_REVENUE                 1275 non-null   float64
 8   TOTAL_EXPENDITURE             1275 non-null   float64
 9   INSTRUCTION_EXPENDITURE       1275 non-null   float64
 10  SUPPORT_SERVICES_EXPENDITURE  1275 non-null   float64
 11  OTHER_EXPENDITURE             1224 non-null   float64
 12  CAPITAL_OUTLAY_EXPENDITURE    1275 non-null   float64
 13  GRA

With previous cell we can see dtypes for all attributes.

In [4]:
df.isnull().sum()*100/df.shape[0]

PRIMARY_KEY                      0.000000
STATE                            0.000000
YEAR                             0.000000
ENROLL                          28.629738
TOTAL_REVENUE                   25.655977
FEDERAL_REVENUE                 25.655977
STATE_REVENUE                   25.655977
LOCAL_REVENUE                   25.655977
TOTAL_EXPENDITURE               25.655977
INSTRUCTION_EXPENDITURE         25.655977
SUPPORT_SERVICES_EXPENDITURE    25.655977
OTHER_EXPENDITURE               28.629738
CAPITAL_OUTLAY_EXPENDITURE      25.655977
GRADES_PK_G                     10.087464
GRADES_KG_G                      4.839650
GRADES_4_G                       4.839650
GRADES_8_G                       4.839650
GRADES_12_G                      4.839650
GRADES_1_8_G                    40.524781
GRADES_9_12_G                   37.551020
GRADES_ALL_G                     4.839650
AVG_MATH_4_SCORE                67.055394
AVG_MATH_8_SCORE                64.897959
AVG_READING_4_SCORE             62

I want to fill just some attributes which has less percent of null value because if the attribute has a high percentage of null value it's not will be wise to fill them with avg for example for 67.1% of "AVG_MATH_4_SCORE" values are null. In this case, if we fill them with avg, avg will not represent the all attribute well.

In [5]:
df["GRADES_PK_G"].fillna(df["GRADES_PK_G"].mean(), inplace=True)
df["GRADES_KG_G"].fillna(df["GRADES_KG_G"].mean(), inplace=True)
df["GRADES_4_G"].fillna(df["GRADES_4_G"].mean(), inplace=True)
df["GRADES_8_G"].fillna(df["GRADES_8_G"].mean(), inplace=True)
df["GRADES_12_G"].fillna(df["GRADES_12_G"].mean(), inplace=True)
df["GRADES_ALL_G"].fillna(df["GRADES_ALL_G"].mean(), inplace=True)

In [6]:
df.isnull().sum()

PRIMARY_KEY                        0
STATE                              0
YEAR                               0
ENROLL                           491
TOTAL_REVENUE                    440
FEDERAL_REVENUE                  440
STATE_REVENUE                    440
LOCAL_REVENUE                    440
TOTAL_EXPENDITURE                440
INSTRUCTION_EXPENDITURE          440
SUPPORT_SERVICES_EXPENDITURE     440
OTHER_EXPENDITURE                491
CAPITAL_OUTLAY_EXPENDITURE       440
GRADES_PK_G                        0
GRADES_KG_G                        0
GRADES_4_G                         0
GRADES_8_G                         0
GRADES_12_G                        0
GRADES_1_8_G                     695
GRADES_9_12_G                    644
GRADES_ALL_G                       0
AVG_MATH_4_SCORE                1150
AVG_MATH_8_SCORE                1113
AVG_READING_4_SCORE             1065
AVG_READING_8_SCORE             1153
dtype: int64

In [7]:
df1 = pd.read_csv("C:/Users/Ahmet/Desktop/data/states_all.csv")

In [8]:
df1.YEAR.unique()

array([1992, 1993, 1994, 1995, 1996, 1997, 1998, 1999, 2000, 2001, 2002,
       2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013,
       2014, 2015, 2016, 1986, 1987, 1988, 1989, 1990, 1991, 2017, 2019],
      dtype=int64)

I will just contniue with one attribute from now on. I picked "GRADES_PK_G" for that.

In [9]:
for i in range(1992,2020):
    df1["GRADES_PK_G"].fillna(df[df1.YEAR == i]["GRADES_PK_G"].mean(),inplace=True)

In [10]:
df2 = pd.read_csv("C:/Users/Ahmet/Desktop/data/states_all.csv")

In [12]:
df2["GRADES_PK_G"] = df2.GRADES_PK_G.interpolate(method='linear',limit_direction="Forward")

In [22]:
print("ortalama ile doldurulan")
print(df["GRADES_PK_G"].describe())
print("\n")
print("yılların kendi içinde ortalaması ile doldurulan")
print(df1["GRADES_PK_G"].describe())
print("\n")
print("interpolasyon yapılan")
print(df2["GRADES_PK_G"].describe())

ortalama ile doldurulan
count      1715.000000
mean      17840.958495
std       28461.478658
min          96.000000
25%        2536.500000
50%       10310.000000
75%       19689.000000
max      256222.000000
Name: GRADES_PK_G, dtype: float64


yılların kendi içinde ortalaması ile doldurulan
count      1715.000000
mean      17103.502244
std       28546.558762
min          96.000000
25%        2536.500000
50%       10310.000000
75%       19689.000000
max      256222.000000
Name: GRADES_PK_G, dtype: float64


interpolasyon yapılan
count      1715.000000
mean      17088.765889
std       29169.489099
min          96.000000
25%        1891.000000
50%        7865.000000
75%       20857.500000
max      256222.000000
Name: GRADES_PK_G, dtype: float64


We filled our null values with 3 methods. As you can see from previous cell, mean, std, median and medians of quartile are different from each other, so we can say that there is remarkable difference among those 3 dataframes.